File: daily/Sliding Window/1234-Replace-the-substring-for-balanced-string.py
class Solution:
    def balancedString(self, s: str) -> int:
        n = len(s)
        limit = n // 4
        outside_frequency = defaultdict(int)
        for c in s:
            outside_frequency[c] += 1
        
        balanced = True
        for c in "QWER":
            if outside_frequency[c] > limit:
               balanced = False
        if balanced: return 0

        res = n
        left = 0
        for right in range(n):
            outside_frequency[s[right]] -= 1
            while left < n:
                is_fixable = True
                for c in "QWER":
                    if outside_frequency[c] > limit:
                        is_fixable = False
                        break
                if is_fixable:
                    res = min(res, right - left + 1)
                    outside_frequency[s[left]] += 1
                    left += 1
                else:
                    break
        return res

File: daily/Sliding Window/1234-Replace-the-substring-for-balanced-string.go
package daily

func balancedString(s string) int {
	n := len(s)
	limit := n / 4
	outsideFrequency := make(map[byte]int)
	for i := 0; i < n; i++ {
		outsideFrequency[s[i]]++
	}

	// Use a string to iterate through the target keys
	targetChars := "QWER"

	isBalanced := true
	for _, char := range targetChars {
		if outsideFrequency[byte(char)] > limit {
			isBalanced = false
			break
		}
	}
	if isBalanced {
		return 0
	}

	res := n
	left := 0
	for right := 0; right < n; right++ {
		outsideFrequency[s[right]]--

		// Manual "all" check
		for left <= right { // Optimized: left shouldn't pass right
			isFixable := true
			for _, char := range targetChars {
				if outsideFrequency[byte(char)] > limit {
					isFixable = false
					break
				}
			}

			if isFixable {
				if right-left+1 < res {
					res = right - left + 1
				}
				outsideFrequency[s[left]]++
				left++
			} else {
				break
			}
		}
	}
	return res
}
